### ISSUE TITLE
dnf backend not honoring "install_weak_deps=False"

### ISSUE THREAD

**Issue by dfaggioli** [PackageKit/PackageKit#486]
As the title suggests, even if `install_weak_deps=False` is present in `/etc/dnf/dnf.conf`, `pkcon install` still tries to install recommended packages.

For example, on openSUSE the QEMU package has the following hard dependencies...
```
$ zypper if --requires qemu
...
Requires       : [8]
    /bin/bash
    group(kvm)
    group(qemu)
    user(qemu)
    /bin/sh
    coreutils
    udev
    acl
```
...and recommended packages:
```
zypper if --recommends qemu
...
Recommends     : [14]
    qemu-tools
    qemu-x86
    qemu-hw-display-virtio-gpu
    qemu-block-curl
    kvm_stat
    qemu-hw-display-qxl
    qemu-hw-display-virtio-gpu-pci
    qemu-hw-display-virtio-vga
    qemu-hw-usb-redirect
    qemu-hw-usb-smartcard
    qemu-ui-curses
    qemu-ui-gtk
    qemu-ui-spice-app
    qemu-ksm = 6.0.0
```
If using, for instance, `microdnf`, things work as expected:
```
$ sudo microdnf install qemu
Package                                                                                                                                                                Repository                                                         Size
Installing:                                                                                                                                                                                                                                   
 acl-2.3.1-1.13.x86_64                                                                                                                                                 repo-oss                                                        76.8 kB
 qemu-6.0.0-24.2.x86_64                                                                                                                                                repo-oss                                                         1.5 MB
 system-user-qemu-20170617-21.13.noarch                                                                                                                                repo-oss                                                        12.7 kB
Transaction Summary:
 Installing:        3 packages
```
If, instead, I try to use `pkcon` (which on this system is configured to use the dnf backend), it wants to install a lot more packages (namely, the recommended ones and their dependencies):
```
$ pkcon install qemu
The following packages have to be installed:
 acl-2.3.1-1.13.x86_64  Commands for Manipulating POSIX Access Control Lists
 kvm_stat-5.12.9-17.16.noarch   Monitoring Tool for KVM guests
 libcacard0-2.8.0-1.6.x86_64    Common Access Card (CAC) emulation
 libefa1-35.0-1.2.x86_64        EFA runtime library
 libfmt7-7.1.3-1.5.x86_64       A formatting library for C++
 libibverbs-35.0-1.2.x86_64     Library & drivers for direct userspace use of InfiniBand/iWARP/RoCE hardware
 libibverbs1-35.0-1.2.x86_64    Ibverbs runtime library
 liblttng-ust0-2.12.0-1.5.x86_64        Linux Trace Toolkit Userspace Tracer library
 libmlx4-1-35.0-1.2.x86_64      MLX4 runtime library
 libmlx5-1-35.0-1.2.x86_64      MLX5 runtime library
 libndctl6-71.1-2.4.x86_64      Management library for "libnvdimm" subsystem devices (Non-volatile Memory)
 libpmem1-1.9-2.4.x86_64        Low-level persistent memory support library
 libpmemobj1-1.9-2.4.x86_64     Persistent Memory Transactional Object Store library
 librados2-16.2.4.26+g555d38aa5a5-1.3.x86_64    RADOS distributed object store client library
 librbd1-16.2.4.26+g555d38aa5a5-1.3.x86_64      RADOS block device client library
 librdmacm1-35.0-1.2.x86_64     Userspace RDMA Connection Manager
 libspice-server1-0.15.0-1.2.x86_64     Library for SPICE-server
 libusbredirparser1-0.9.0-1.3.x86_64    A protocol for redirecting USB traffic - Client-side library
 libvirglrenderer1-0.9.1-1.2.x86_64     Virgil3D renderer
 pmdk-1.9-2.4.x86_64    Persistent Memory Development Kit
 python38-curses-3.8.10-1.2.x86_64      Python Interface to the (N)Curses Library
 qemu-6.0.0-24.2.x86_64 Machine emulator and virtualizer
 qemu-audio-spice-6.0.0-24.2.x86_64     Spice based audio support for QEMU
 qemu-block-curl-6.0.0-24.2.x86_64      cURL block support for QEMU
 qemu-block-rbd-6.0.0-24.2.x86_64       Rados Block Device (Ceph) support for QEMU
 qemu-chardev-spice-6.0.0-24.2.x86_64   Spice vmc and port chardev support for QEMU
 qemu-hw-display-qxl-6.0.0-24.2.x86_64  QXL display support for QEMU
 qemu-hw-display-virtio-gpu-6.0.0-24.2.x86_64   Virtio GPU display support for QEMU
 qemu-hw-display-virtio-gpu-pci-6.0.0-24.2.x86_64       Virtio-gpu pci device for QEMU
 qemu-hw-display-virtio-vga-6.0.0-24.2.x86_64   Virtio vga device for QEMU
 qemu-hw-usb-redirect-6.0.0-24.2.x86_64 USB redirection support for QEMU
 qemu-hw-usb-smartcard-6.0.0-24.2.x86_64        USB smartcard support for QEMU
 qemu-ksm-6.0.0-24.2.x86_64     Kernel Samepage Merging services
 qemu-tools-6.0.0-24.2.x86_64   Tools for QEMU
 qemu-ui-curses-6.0.0-24.2.x86_64       Curses based UI support for QEMU
 qemu-ui-gtk-6.0.0-24.2.x86_64  GTK based UI support for QEMU
 qemu-ui-opengl-6.0.0-24.2.x86_64       OpenGL based UI support for QEMU
 qemu-ui-spice-app-6.0.0-24.2.x86_64    Spice UI support for QEMU
 qemu-ui-spice-core-6.0.0-24.2.x86_64   Core Spice support for QEMU
 rdma-core-35.0-1.2.x86_64      RDMA core userspace libraries and daemons
 system-user-qemu-20170617-21.13.noarch System user and group qemu
```
Which I think tells us that the fact that we have `install_weak_deps=False` is basically being ignored

**Comment by dfaggioli** [comment]
I shall add that @Conan-Kudo suggested that maybe this might be dealt with at the libdnf level, which is why we created this: https://bugzilla.redhat.com/show_bug.cgi?id=1955484

**Comment by dfaggioli** [comment]
> I shall add that @Conan-Kudo suggested that maybe this might be dealt with at the libdnf level, which is why we created this: https://bugzilla.redhat.com/show_bug.cgi?id=1955484

However, since not much is happening on that front, I gave it a quick try myself. Basing on what I saw here: https://github.com/rpm-software-management/microdnf/blob/7b84c0699311839ea79b0cfbcbf88ef91e3c0ce1/dnf/plugins/install/dnf-command-install.c#L81-L82

I created this patch: https://build.opensuse.org/package/view_file/home:dfaggioli:microos-desktop/PackageKit/PackageKit-dnf-ignore-weak-deps.patch?expand=1 which, however, does not appear to be working for me.

Note that the patch is there in a downstream package only because it was the quickest way for me to test it... I'm happy to work with and submit it upstream if it is interesting (and if someone helps me understand why it does not work and how to do it right :-D)

**Comment by Conan-Kudo** [comment]
Please feel free to submit a pull request, @hughsie, myself, and DNF folks can review it and help out. 😉

**Comment by dfaggioli** [comment]
> Please feel free to submit a pull request! @hughsie, myself, and DNF folks can review it and help out.

Oh, sure, thanks! Problem is, the patch I have is not working. That is, with it applied, `pkcon` still install the full set of recommended packages (and their respective dependencies). And I'm not so sure about submitting a pull request for a non-working change... Or shall I ? :-)

If anyone stop what I'm doing wrong there (I looked at PackageKit code for the first time today, so there's plenty of chanches that I've made some glaring mistakes), I'd be happy to continue developing and testing it.

I can attach it (is it even possible?) or paste it in a comment if that helps...

**Comment by Conan-Kudo** [comment]
@dfaggioli Just make a draft PR and we can take a look and see what's going wrong!

**Comment by dfaggioli** [comment]
> @dfaggioli Just make a draft PR and we can take a look and see what's going wrong!

Ok then. Too late now (sorry), but I'll try to do that tomorrow morning. Thanks again!

**Comment by dfaggioli** [comment]
> @dfaggioli Just make a draft PR and we can take a look and see what's going wrong!

ok, while looking better/doing more debugging, I think I've understood that I probably need to "act" at the level of the `dnf_utils_create_sack_for_filters()` function, rather than inside `pk_backend_search_thread()`.

I'll look further (but of course, any suggestion is more welcome :-D )

**Comment by dfaggioli** [comment]
> > @dfaggioli Just make a draft PR and we can take a look and see what's going wrong!
> 
> ok, while looking better/doing more debugging, I think I've understood that I probably need to "act" at the level of the `dnf_utils_create_sack_for_filters()` function, rather than inside `pk_backend_search_thread()`.
> 
Well, not really... Probably it's much better to fiddle with `pk_backend_transaction_run()`, i.e., by calling `dnf_transaction_set_dont_solve_goal()` inside it (among other things)...

**Comment by dfaggioli** [comment]
> > > @dfaggioli Just make a draft PR and we can take a look and see what's going wrong!
> > 
> > 
> > ok, while looking better/doing more debugging, I think I've understood that I probably need to "act" at the level of the `dnf_utils_create_sack_for_filters()` function, rather than inside `pk_backend_search_thread()`.
> 
> Well, not really... Probably it's much better to fiddle with `pk_backend_transaction_run()`, i.e., by calling `dnf_transaction_set_dont_solve_goal()` inside it (among other things)...

And I think I have it working now... Let's see if I can put together a pull request.

BTW, yes, I know, that I've been talking to myself all the time.

But I'm not crazy.

Or at least I don't think I am. Oh, well... :-O

**Comment by dfaggioli** [comment]
> @dfaggioli Just make a draft PR and we can take a look and see what's going wrong!

@Conan-Kudo done! https://github.com/PackageKit/PackageKit/pull/488 :-)

It's not a draft because it now works for me, and I think it's actually right. Of course, I still have close to zero experience with PackageKit and libdnf codebase, so I can be totally wrong, and I'm happy to learn in what ways and try to fix things.

**Comment by Conan-Kudo** [comment]
Fixed with ecd4a969939855350930511516ae20584bda7fa7
